src/xrpld/rpc/handlers/Simulate.cpp

```diff
@@ -0,0 +1,307 @@
+            return feeOrError;
+        tx_json[jss::Fee] = feeOrError;
+    }
+
+    if (!tx_json.isMember(sfSigningPubKey.jsonName))
+    {
+        // autofill SigningPubKey
+        tx_json[sfSigningPubKey.jsonName] = "";
+    }
+
+    if (tx_json.isMember(jss::Signers))
+    {
+        if (!tx_json[jss::Signers].isArray())
+            return RPC::invalid_field_error("tx.Signers");
+        // check multisigned signers
+        for (int index = 0; index < tx_json[jss::Signers].size(); index++)
+        {
+            auto& signer = tx_json[jss::Signers][index];
+            if (!signer.isObject() || !signer.isMember(jss::Signer) ||
+                !signer[jss::Signer].isObject())
+                return RPC::invalid_field_error(
+                    "tx.Signers[" + std::to_string(index) + "]");
+            if (!signer[jss::Signer].isMember(sfTxnSignature.jsonName))
+            {
+                // autofill TxnSignature
+                signer[jss::Signer][sfTxnSignature.jsonName] = "";
+            }
+            else if (signer[jss::Signer][sfTxnSignature.jsonName] != "")
+            {
+                // Transaction must not be signed
+                return rpcError(rpcTX_SIGNED);
+            }
+        }
+    }
+
+    if (!tx_json.isMember(sfTxnSignature.jsonName))
+    {
+        // autofill TxnSignature
+        tx_json[sfTxnSignature.jsonName] = "";
+    }
+    else if (tx_json[sfTxnSignature.jsonName] != "")
+    {
+        // Transaction must not be signed
+        return rpcError(rpcTX_SIGNED);
+    }
+
+    if (!tx_json.isMember(jss::Sequence))
+    {
+        // autofill Sequence
+        bool const hasTicketSeq = tx_json.isMember(sfTicketSequence.jsonName);
+        auto const accountStr = tx_json[jss::Account];
+        if (!accountStr.isString())
+        {
+            // sanity check, should fail earlier
+            return RPC::invalid_field_error("tx.Account");  // LCOV_EXCL_LINE
+        }
+
+        auto const srcAddressID =
+            parseBase58<AccountID>(tx_json[jss::Account].asString());
+        if (!srcAddressID.has_value())
+        {
+            return RPC::make_error(
+                rpcSRC_ACT_MALFORMED, RPC::invalid_field_message("tx.Account"));
+        }
+
+        std::shared_ptr<SLE const> sle =
+            context.app.openLedger().current()->read(
+                keylet::account(*srcAddressID));
+        if (!hasTicketSeq && !sle)
+        {
+            JLOG(context.app.journal("Simulate").debug())
+                << "Failed to find source account "
+                << "in current ledger: " << toBase58(*srcAddressID);
+
+            return rpcError(rpcSRC_ACT_NOT_FOUND);
+        }
+
+        tx_json[jss::Sequence] = hasTicketSeq
+            ? 0
+            : context.app.getTxQ().nextQueuableSeq(sle).value();
+    }
+
+    return std::nullopt;
+}
+
+// {
+//   tx_blob: <string> XOR tx_json: <object>,
+//   binary: <bool>
+// }
+Json::Value
+doSimulate(RPC::JsonContext& context)
```

Comment:
This function is too big, please divide it.

Reply:
I tried making some improvements, but it only made it more convoluted.

Reply:
Thanks! Fixed in 5b42a96fec72ed3246f6073445918aad69efcb36